include/exec/__detail/intrusive_heap.hpp

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) 2024 Maikel Nadolski
+ * Copyright (c) 2024 NVIDIA Corporation
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "../../stdexec/__detail/__config.hpp"
+
+#include <cstddef>
+#include <bit>
+
+namespace exec {
+#if defined(__cpp_lib_int_pow2) && __cpp_lib_int_pow2 >= 202002L
+  namespace detail {
+    using std::bit_ceil;
+  } // namespace detail
+#else
+#  if defined(__has_builtin) && __has_builtin(__builtin_clzll)
+  namespace detail {
+    inline std::size_t bit_ceil(std::size_t n) noexcept {
+      const int leading_zeros = __builtin_clzll(n);
+      return std::size_t{1} << (sizeof(std::size_t) * 8 - leading_zeros + 1);
+    }
+  } // namespace detail
+#  else
+  namespace detail {
+    inline std::size_t bit_ceil(std::size_t n) noexcept {
+      std::size_t m = n - 1;
+      m |= m >> 1;
+      m |= m >> 2;
+      m |= m >> 4;
+      m |= m >> 8;
+      m |= m >> 16;
+      m |= m >> 32;
+      return m + 1;
+    }
+  } // namespace detail
+#  endif
+#endif
+
+  template <auto Key, auto Prev, auto Left, auto Right>
+  class intrusive_heap;
```

Comment:
I apologize for the invasion, but I worked on a very similar thing and came to the thought, that erasures from heap are too expensive and therefore cancelation is also. Isn't it better to implement intrusive ordered set based on any balanced binary tree? To my shame I never implemented this idea, but maybe you can find it interesting.

Reply:
I think erasure here is O(log N) (I set the time point value to 0 and call bottom-up-heapify). Thats pretty good.

For a real production version of this I have seen that @lewissbaker or @kirkshoop mentioned some hash based priority queue. But I don't know how that works.

Reply:
Yes, you're right, I was too caught up with the thought about multiple erasures that time, sorry

Reply:
Note that this is not a generic ordered set. Deleting an element modifies the key value which is unusual. More over it assumes that a zero initialized key value yields a minimum. More over bottom up heapify will pass the node up even if the parent has a zero timeout, too. That way it is ensured that after calling bottom up the node to be erased is at root and can be popped back.

This is a data structure that works for timeouts but is probably not good for a generic setting